### PR TITLE
M4 conan v2

### PR DESCRIPTION
**Pull request by jellespijker**
See PR in CCI: https://github.com/conan-io/conan-center-index/pull/12815

This PR and recipe is purely for use in the Ultimaker Cura workflow, and will be used until the above PR is merged into CCI

Fixes: CURA-9595
Contributes to: CURA-9574, CURA-9575, CURA-9578, CURA-9587, CURA-9571
Is depend on: CURA-9573

### CODE REVIEW

File: recipes/m4/all/conanfile.py [hunk 1]
```diff
@@ -1,9 +1,100 @@
...
+    from conan.tools.env.environment import environment_wrap_command
+    """ Will wrap a unix command inside a bash terminal It requires to have MSYS2, CYGWIN, or WSL"""
+
+    subsystem = conanfile.conf.get("tools.microsoft.bash:subsystem")
+    shell_path = conanfile.conf.get("tools.microsoft.bash:path")
+    if not subsystem or not shell_path:
+        raise ConanException("The config 'tools.microsoft.bash:subsystem' and "
+                             "'tools.microsoft.bash:path' are "
+                             "needed to run commands in a Windows subsystem")
+    env = env or []
```
Comment (casperlamboo) [inline recipes/m4/all/conanfile.py:34]: Slightly prefer
```suggestion
    if !env:
        env = []
```

File: recipes/m4/all/conanfile.py [hunk 2]
```diff
@@ -1,9 +1,100 @@
...
+    if subsystem == MSYS2:
+        # Configure MSYS2 to inherith the PATH
+        msys2_mode_env = Environment()
+        _msystem = {"x86": "MINGW32"}.get(conanfile.settings.get_safe("arch"), "MINGW64")
+        msys2_mode_env.define("MSYSTEM", _msystem)
```
Comment (casperlamboo) [inline recipes/m4/all/conanfile.py:39]: Why prepend and underscore with the variable `_msystem`? For me this indicates either a private variable, or a variable that is not used. Neither is true in this case.
```suggestion
        msystem = {"x86": "MINGW32"}.get(conanfile.settings.get_safe("arch"), "MINGW64")
        msys2_mode_env.define("MSYSTEM", msystem)
```

File: recipes/m4/all/conanfile.py [hunk 3]
```diff
@@ -1,9 +1,100 @@
...
+    def configure(self, build_script_folder=None, args=None):
+        """
+        http://jingfenghanmax.blogspot.com.es/2010/09/configure-with-host-target-and-build.html
+        https://gcc.gnu.org/onlinedocs/gccint/Configure-Terms.html
+        """
+        script_folder = path.join(self._conanfile.source_folder, build_script_folder) \
+            if build_script_folder else self._conanfile.source_folder
+
+        configure_args = []
+        configure_args.extend(args or [])
```
Comment (casperlamboo) [inline recipes/m4/all/conanfile.py:75]:
```suggestion
        if args:
            configure_args.extend(args)
```

File: recipes/m4/all/conanfile.py [hunk 4]
```diff
@@ -1,9 +1,100 @@
...
+        make_program = self._conanfile.conf.get("tools.gnu:make_program",
+                                                default="mingw32-make" if self._use_win_mingw() else "make")
+        str_args = self._make_args
+        str_extra_args = " ".join(args) if args is not None else ""
+        jobs = ""
+        if "-j" not in str_args and "nmake" not in make_program.lower():
+            njobs = build_jobs(self._conanfile)
+            if njobs:
+                jobs = "-j{}".format(njobs)
+        command = join_arguments([make_program, target, str_args, str_extra_args, jobs])
```
Comment (casperlamboo) [inline recipes/m4/all/conanfile.py:96]: I get why you do it like this; it is nice to see a list of all arguments applied to the command. But as there are various optional arguments i prefer something like.
```suggestion
        command = [make_program, targert, self._make_args]
        if args is not None:
            command.append(" ".join(args))
        if "-j" not in str_args and "nmake" not in make_program.lower():
            njobs = build_jobs(self._conanfile)
            if njobs:
                jobs = "-j{}".format(njobs)
                command.append(jobs)
        command = join_arguments(command)
```
Please check my logic. This was written in a githubs web-editor

File: recipes/m4/all/conanfile.py [hunk 5]
```diff
@@ -14,117 +105,106 @@ class M4Conan(ConanFile):
...
+        if is_msvc(self):
+            env.define("AR", f"{unix_path(self, self.source_folder)}/build-aux/ar-lib lib")
+            env.define("LD", "link")  # otherwise configure reports "ld" as its linker. See conan-io/conan#11922
+            env.define("NM", "dumpbin -symbols")
+            env.define("OBJDUMP", ":")
+            env.define("RANLIB", ":")
+            env.define("STRIP", ":")
+        at.generate(env)
+
+    @lru_cache(1)
```
Comment (casperlamboo) [inline recipes/m4/all/conanfile.py:173]: If I understand correctly we are caching the autotools at the build step. This way when retrieving the cache in the packaging cache we get the autotools generetate from a the state where it was in the build step?

In my opinion this doesn't make the data flow very clear. I prefer to set a `self._autotools` property in the build step that is read in the packaging step; it is less fancy but the flow is more clear.

Comment (casperlamboo) [reply recipes/m4/all/conanfile.py:173]: On the other hand, this makes it possible to perform the packaging step independent from the build step which is nice for debugging.

File: recipes/m4/all/conanfile.py [hunk 6]
```diff
@@ -14,117 +105,106 @@ class M4Conan(ConanFile):
...
     def package(self):
-        self.copy("COPYING", src=self._source_subfolder, dst="licenses")
-        with self._build_context():
-            autotools = self._configure_autotools()
-            autotools.install()
-        tools.rmdir(os.path.join(self.package_folder, "share"))
+        autotools = self._autotools()
+        # KB-H013 we're packaging an application, place everything under bin
+        autotools.install(args=[f"DESTDIR={unix_path(self, path.join(self.package_folder, 'bin'))}"])
+        copy(self, "COPYING*", src=self.source_folder, dst=path.join(self.package_folder, "licenses"))
```
Comment (casperlamboo) [inline recipes/m4/all/conanfile.py:190]:
```suggestion
        licence_folder = path.join(self.package_folder, "licenses")
        copy(self, "COPYING*", src=self.source_folder, dst=licence_folder)
```

**Comment by jellespijker** [comment]
Shot sorry @casperlamboo something went wrong here. You seemed to have reviewed an old branch.

https://github.com/Ultimaker/conan-center-index/tree/CURA-9595_m4_conan_v2

**Pull request closed without merging**